cub/test/catch2_test_device_segmented_reduce_iterators_64bit.cu

```diff
@@ -0,0 +1,152 @@
+#include "catch2/catch.hpp"
+#include "catch2_test_cdp_helper.h"
+#include "catch2_test_helper.h"
+
+DECLARE_CDP_WRAPPER(cub::DeviceSegmentedReduce::Reduce, device_segmented_reduce);
+DECLARE_CDP_WRAPPER(cub::DeviceSegmentedReduce::Sum, device_segmented_sum);
+
+// %PARAM% TEST_CDP cdp 0:1
+
+// List of types to test
+using custom_t           = std::size_t; //c2h::custom_type_t<c2h::accumulateable_t, c2h::equal_comparable_t>;
+using iterator_type_list = c2h::type_list<type_pair<custom_t>, type_pair<std::size_t>>;
+
+// multiply by constant
+template <typename T>
+class cmult {
+    public:
+    __host__ __device__
+    constexpr cmult(T mult) : mult(mult) {}
+    __host__ __device__
+    constexpr T operator()(T const& value) const {
+        return value*mult;
+    }
+    private:
+    T mult;
+};
+
+CUB_TEST("Device segmented reduce works with fancy input iterators and 64-bit offsets",
+         "[reduce][device]",
+         iterator_type_list)
+{
+  using params   = params_t<TestType>;
+  using item_t   = typename params::item_t;
+  using output_t = typename params::output_t;
+  using offset_t = std::ptrdiff_t;
+
+  constexpr offset_t min_items_rand = static_cast<offset_t>(0);
+  constexpr offset_t max_items_rand = static_cast<offset_t>(1) << 7;
+  constexpr offset_t offset_base_value = static_cast<offset_t>(1) << 32;
+  constexpr int max_segments = 4;
+
+  // Number of items
+  const offset_t num_items_rand = GENERATE_COPY(take(2, random(min_items_rand, max_items_rand)),
+                                           values({
+                                             min_items_rand,
+                                             max_items_rand,
+                                           }));
+  INFO("Test num_items_rand: " << num_items_rand);
+
+  // Range of segment sizes to generate
+  const std::tuple<offset_t, offset_t> seg_size_range =
+    GENERATE_COPY(table<offset_t, offset_t>({{1, 1}, {1, max_segments}, {max_segments, max_segments}}));
+  INFO("Test seg_size_range: [" << std::get<0>(seg_size_range) << ", "
+                                << std::get<1>(seg_size_range) << "]");
+
+  // Generate randomized offsets that will be added to 2^32
+  thrust::device_vector<offset_t> segment_offsets =
+    c2h::gen_uniform_offsets<offset_t>(CUB_SEED(1),
+                                       num_items_rand,
+                                       std::get<0>(seg_size_range),
+                                       std::get<1>(seg_size_range));
+  const int num_segments = static_cast<int>(segment_offsets.size() - 1);
+  // Add 2^32 to offsets
+  thrust::transform(segment_offsets.begin(),
+                    segment_offsets.end(),
+                    thrust::make_transform_iterator(
+                      thrust::make_counting_iterator(static_cast<offset_t>(0)),
+                      cmult<offset_t>(offset_base_value)
+                    ),
+                    segment_offsets.begin(),
+                    thrust::plus<offset_t>{});
```

Comment:
suggestion: the test seems to be a bit complicated, do you think we might go with something like the code below?
```suggestion
  constexpr offset_t min_items = 1ull << 31;
  constexpr offset_t max_items = 1ull << 33;
  constexpr offset_t offset_base_value = static_cast<offset_t>(1) << 32;

  // Number of items
  const offset_t num_items_in_first_segment = GENERATE_COPY(take(2, random(min_items, max_items)));
  const offset_t num_items_in_second_segment = GENERATE_COPY(take(2, random(min_items, max_items)));

  thrust::device_vector<offset_t> segment_offsets = {0ull,
                                                     num_items_in_first_segment,
                                                     num_items_in_first_segment +
                                                       num_items_in_second_segment};
```